contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
 
 ##### Typical status properties
 
-**Conditions** represent the latest available observations of an object's
-state.  They are an extension mechanism intended to be used when the details of
-an observation are not a priori known or would not apply to all instances of a
-given Kind.  For observations that are well known and apply to all instances, a
-regular field is preferred.  An example of a Condition that probably should
-have been a regular field is Pod's "Ready" condition - it is managed by core
-controllers, it is well understood, and it applies to all Pods.
+**Conditions** provide a standard mechanism for higher-level status reporting
+from a controller. They are an extension mechanism which allows tools and other
+controllers to collect summary information about resources without needing to
+understand resource-specific status details. Conditions should complement more
+detailed information about the observed status of an object written by a
+controller, rather than replace it. For example, the "Available" condition of a
+Deployment can be determined by examining `readyReplicas`, `replicas`, and
+other properties of the Deployment. However, the "Available" condition allows
+other components to avoid duplicating the availability logic in the Deployment
+controller.
 
 Objects may report multiple conditions, and new types of conditions may be
 added in the future or by 3rd party controllers. Therefore, conditions are
-represented using a list/slice, where all have similar structure.
+represented using a list/slice of objects, where each condition has a similar
+structure. Conditions are most useful when they follow some consistent
+conventions:
+
+* Conditions should be added to explicitly convey properties that users and
+  components care about rather than requiring those properties to be inferred
+  from other observations.  Once defined, the meaning of a Condition can not be
+  changed arbitrarily - it becomes part of the API, and has the same backwards-
+  and forwards-compatibility concerns of any other part of the API.
+
+* Within a resource, Conditions should have a consistent polarity. **A
+  "positive" polarity where "True" indicates normal operation and "False"
+  indicates a failure is recommended if there is not an established polarity.**
+  (Note that this is an explicit reversal from earlier recommendations of
```

Comment:
Have we surveyed all of the built-in conditions for this?

Just at a glance:
* Node has Ready (positive) and *Pressure and NetworkUnavailable (all negative)
* Namespace has *Failure (negative) and *Remaining (ambiguous)
* PVC has Resizing (negative)
* Pod has all positive
* RC has failure
* Deployment, HPA are mixed

Reply:
Looking through GoDoc at https://godoc.org/k8s.io/api :
* At least 12 packages have nothing exported matching "ConditionType"

* HPA's "ScalingLimited" is not clearly a failure, but might be an exceptional case.
* Job has "Complete" and "Failed" (and "Succeeded" is *Complete=True and Failed unset*)
* ReplicaSet has only "ReplicaFailure" (mirroring ReplicationController, and also mirrored into Deployment), and DaemonSet and StatefulSet have no actual Condition types, despite defining types (!)
* PriorityLevel and FlowSchema also define a type but no conditions
* PVC's "Resizing" isn't positive/negative -- it's a non-disruptive indication of work in progress (and failure to expand the volume isn't indicated by Condition at all)

I think it's fair to characterize this as "all over the place".

Unfortunately, my experience is that lacking a consistent "is this working" / "what's not working" mechanism makes Kubernetes less approachable for newcomers, who end up needing to cram the entire system into their head when debugging (or throw up their hands and find someone who has already done so).

Reply:
I agree that's a defect, but I don't think better defining conditions is going to solve it.

I think it'd be fun to write a "quiescence controller" that watches existing objects, and maintains a "quiescent" condition. Then we'd just have to write that logic in one place.

Reply:
If we want to say that there's a convention on polarity, we should commit to going through EVERY SINGLE built-in condition, deprecating the ones that are "wrong" (can't delete, but...) and adding an equivalent-opposite in the "correct" polarity.

Reply:
I don't like the polarity convention, except that conditions must summarize to both humans and machines, and the polarity should make sense to a human in the context of the condition type, not as part of some aribrarty rule.

NotReady = true is strictly harder for a human to understand than Ready = false and for that reason alone should be excluded.  We choose to name fields by what they do, rather than by how they are introduced - why would we have different rules for conditions?

Reply:
I agree w/ Clayton; we probably got into this situation by not having a clear understanding that a condition not being present means "unknown" not "false".

Unfortunately, "polarity is for people" means that a aggregating controller like I talk about above needs an extra bit of information, but I think that's still tractable.

Reply:
This seems like the most contentious part of this PR.   Would the author like to replace the proposed changes to "polarity" with something like:

````For some conditions, True represents normal operation, and for some "False" represents normal operation.  It is not possible to compute an overall "summary" condition from all the conditions in an object in a a generic way.```

`without detailed knowledge of the specific  are Tpull out the polarity changes, or maybe just change the existing language about polarity to acknowledge that